packages/wasm/crate/src/storage.rs

```diff
@@ -104,39 +106,48 @@ impl IndexedDBStorage {
             .transaction_on_one(&self.constants.tables.spendable_notes)?;
         let store = idb_tx.object_store(&self.constants.tables.spendable_notes)?;
 
+        let asset_id: Option<asset::Id> = request.asset_id.map(TryInto::try_into).transpose()?;
+        let address_index: Option<AddressIndex> =
+            request.address_index.map(TryInto::try_into).transpose()?;
+        let amount_to_spend: Option<Amount> =
+            request.amount_to_spend.map(TryInto::try_into).transpose()?;
+
+        if let (None, Some(_)) = (asset_id, amount_to_spend) {
+            return Err(
+                anyhow::anyhow!("specified amount_to_spend without asset_id filter").into(),
+            );
+        }
+
+        let mut records = Vec::new();
+        let mut total = Amount::zero();
+
         let raw_values = store.get_all()?.await?;
-        let parsed_notes = raw_values
-            .into_iter()
-            .filter_map(|js_value| self.parse_note(js_value, &request).ok())
-            .collect();
+        for raw_record in raw_values {
+            let record: SpendableNoteRecord = serde_wasm_bindgen::from_value(raw_record)?;
 
-        Ok(parsed_notes)
-    }
+            if !request.include_spent && record.height_spent.is_some() {
+                continue;
+            }
 
-    fn parse_note(
-        &self,
-        js_value: JsValue,
-        request: &NotesRequest,
-    ) -> WasmResult<SpendableNoteRecord> {
-        let note: SpendableNoteRecord = serde_wasm_bindgen::from_value(js_value)?;
-
-        let asset_id_matches = match &request.asset_id {
-            Some(asset_id) => note.note.asset_id() == asset_id.clone().try_into()?,
-            None => true,
-        };
-
-        let address_index_matches = match &request.address_index {
-            Some(address_index) => note.address_index.eq(&address_index.clone().try_into()?),
-            None => true,
-        };
-
-        if asset_id_matches && address_index_matches && note.height_spent.is_none() {
-            Ok(note)
-        } else {
-            Err(WasmError::Anyhow(anyhow::anyhow!(
-                "Note does not match the request"
-            )))
+            if Some(record.note.asset_id()) != asset_id {
+                continue;
+            }
+
+            if Some(record.address_index) != address_index {
+                continue;
+            }
+
+            total += record.note.amount();
+            records.push(record);
+
+            if let Some(amount_to_spend) = amount_to_spend {
+                if total >= amount_to_spend {
+                    break;
+                }
+            }
```

Comment:
This fixes a bug in the previous implementation -- if you've ever looked at a transaction and noticed that it spent more notes than necessary, this is why. (The behavior is however no longer used by the current planner implementation).